packages/xstate-wallet/src/browser-wallet.ts

```diff
@@ -0,0 +1,187 @@
+type OnNewMessage = (message: Message) => void;
+
+type DepositInfo = {
+  depositAt: Uint256;
+  myDeposit: Uint256;
+  totalAfterDeposit: Uint256;
+  fundedAt: Uint256;
+};
+
+export class BrowserWallet {
+  private constructor(
+    private onNewMessage: OnNewMessage,
+    private chain = new ChainWatcher(),
+    public store = new Store(chain),
+    private registeredChannels = new Set<string>(),
+    // TODO: probably should be stored in a more permanent storage
+    private channelFunding: Dictionary<Funding> = {}
+  ) {}
+
+  private async init(): Promise<BrowserWallet> {
+    await this.store.initialize();
+    return this;
+  }
+
+  static async create(onNewMessage: OnNewMessage): Promise<BrowserWallet> {
+    return new BrowserWallet(onNewMessage).init();
+  }
+
+  async incomingMessage(payload: Message): Promise<Message> {
+    let response: Message = {
+      objectives: [],
+      signedStates: []
+    };
+    // Store any new objectives
+    const payloadObjective = payload.objectives?.[0];
+    if (!payloadObjective) {
+      logger.info('No incoming objectives');
+    } else {
+      await this.store.addObjective(payloadObjective);
+    }
+
+    // Store any new states
+    const payloadState = payload.signedStates?.[0];
+    if (!payloadState) {
+      logger.info('No incoming states');
+    } else {
+      await this.store.addState(payloadState);
+      const channelId = calculateChannelId(payloadState);
+      if (!this.registeredChannels.has(channelId)) {
+        this.chain.chainUpdatedFeed(channelId).subscribe({
+          next: chainInfo => this.onFundingUpdate(channelId, chainInfo)
+        });
+      }
+    }
+
+    // Crank objectives
+    for (const objective of this.store.objectives) {
+      switch (objective.type) {
+        case 'OpenChannel': {
+          response = await this.onOpenChannelObjective(objective.data.targetChannelId);
+          break;
+        }
+        default:
+          throw new Error('Objective not supported');
+      }
+    }
+    return response;
+  }
+
+  async onOpenChannelObjective(channelId: string): Promise<Message> {
+    const channel = await this.store.getEntry(channelId);
+    const pk = await this.store.getPrivateKey(await this.store.getAddress());
+    const depositInfo = await this.getDepositInfo(channelId);
+
+    const response = this.crankOpenChannelObjective(
+      channel,
+      this.channelFunding[channelId],
+      depositInfo,
+      pk
+    );
+    if (response.deposit) {
+      this.channelFunding[channelId] = {...this.channelFunding[channelId], status: 'DEPOSITED'};
+      await this.chain.deposit(
+        channel.channelId,
+        this.channelFunding[channelId].amountOnChain,
+        depositInfo.myDeposit
+      );
+    }
+    if (response.signedStates[0]) {
+      await this.store.addState(response.signedStates[0]);
+      this.onNewMessage(response);
+    }
+    return response;
+  }
+
+  crankOpenChannelObjective(
```

Comment:
If this is meant to be a pure function, is there a reason to have it be a (public) method on a class, besides convenience?

Reply:
https://github.com/statechannels/statechannels/pull/3402#discussion_r603335387